vms/evm/sync/customrawdb/sync_progress.go

```diff
@@ -0,0 +1,235 @@
+	// and is equal to [syncPerformedPrefix] + block number as uint64.
+	syncPerformedKeyLength = len(syncPerformedPrefix) + wrappers.LongLen
+)
+
+// ReadSyncRoot reads the root corresponding to the main trie of an in-progress
+// sync and returns common.Hash{} if no in-progress sync was found.
+func ReadSyncRoot(db ethdb.KeyValueReader) (common.Hash, error) {
+	ok, err := db.Has(syncRootKey)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	if !ok {
+		return common.Hash{}, database.ErrNotFound
+	}
+	root, err := db.Get(syncRootKey)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	return common.BytesToHash(root), nil
+}
+
+// WriteSyncRoot writes root as the root of the main trie of the in-progress sync.
+func WriteSyncRoot(db ethdb.KeyValueWriter, root common.Hash) error {
+	return db.Put(syncRootKey, root[:])
+}
+
+// WriteCodeToFetch adds a marker that we need to fetch the code for `hash`.
+func WriteCodeToFetch(db ethdb.KeyValueWriter, codeHash common.Hash) error {
+	return db.Put(codeToFetchKey(codeHash), nil)
+}
+
+// DeleteCodeToFetch removes the marker that the code corresponding to `hash` needs to be fetched.
+func DeleteCodeToFetch(db ethdb.KeyValueWriter, codeHash common.Hash) error {
+	return db.Delete(codeToFetchKey(codeHash))
+}
+
+// NewCodeToFetchIterator returns a KeyLength iterator over all code
+// hashes that are pending syncing. It is the caller's responsibility to
+// parse the key and call Release on the returned iterator.
+func NewCodeToFetchIterator(db ethdb.Iteratee) ethdb.Iterator {
+	return rawdb.NewKeyLengthIterator(
+		db.NewIterator(CodeToFetchPrefix, nil),
+		codeToFetchKeyLength,
+	)
+}
+
+func codeToFetchKey(codeHash common.Hash) []byte {
+	codeToFetchKey := make([]byte, codeToFetchKeyLength)
+	copy(codeToFetchKey, CodeToFetchPrefix)
+	copy(codeToFetchKey[len(CodeToFetchPrefix):], codeHash[:])
+	return codeToFetchKey
+}
+
+// NewSyncSegmentsIterator returns a KeyLength iterator over all trie segments
+// added for root. It is the caller's responsibility to parse the key and call
+// Release on the returned iterator.
+func NewSyncSegmentsIterator(db ethdb.Iteratee, root common.Hash) ethdb.Iterator {
+	segmentsPrefix := make([]byte, len(syncSegmentsPrefix)+common.HashLength)
+	copy(segmentsPrefix, syncSegmentsPrefix)
+	copy(segmentsPrefix[len(syncSegmentsPrefix):], root[:])
+
+	return rawdb.NewKeyLengthIterator(
+		db.NewIterator(segmentsPrefix, nil),
+		syncSegmentsKeyLength,
+	)
+}
+
+// WriteSyncSegment adds a trie segment for root at the given start position.
+func WriteSyncSegment(db ethdb.KeyValueWriter, root common.Hash, start common.Hash) error {
+	// packs root and account into a key for storage in db.
+	bytes := make([]byte, syncSegmentsKeyLength)
+	copy(bytes, syncSegmentsPrefix)
+	copy(bytes[len(syncSegmentsPrefix):], root[:])
+	copy(bytes[len(syncSegmentsPrefix)+common.HashLength:], start.Bytes())
+	return db.Put(bytes, []byte{0x01})
```

Comment:
Why do we write a `[]byte{0x01}` here?
```suggestion
	return db.Put(bytes, nil)
```